### ISSUE TITLE
Plotly offline is asking for credentials.

### ISSUE THREAD

**Issue by twocs** [imatlab/imatlab#13]
Plotly offline is asking for credentials. This is listed as a longstanding bug. https://github.com/plotly/MATLAB-Online/issues/75

Basically, running the command `fig2plotly(gcf, 'offline', true);` calls `plotlyfig`, which calls `signin`, which loads the config from `~/.plotly/.config` _before_ the 'offline' arg is checked.
```
>> fig2plotly(gcf, 'offline', true);
Error using plotlyfig
(line 47)
 
Oops! You must be signed
in to initialize a
plotlyfig object.
```

It seems that the config file isn't set up in a way that lets me proceed without credentials. So my question is what should be done to set up the plotly config file?

**Comment by anntzer** [comment]
Did you follow the instructions at https://github.com/imatlab/imatlab#inline-graphics?

**Comment by twocs** [comment]
Yes, i did all those steps. 

My thinking is that setting the config for plotly matlab must be performed once. But what has everyone else been doing to set the values for these files?

**Comment by imatlab-helper** [comment]
This is curious an previously unobserved.

You successfully called ```getplotlyoffline('https://cdn.plot.ly/plotly-latest.min.js')``` from the matlab notebook?
What happens if you just call ```plot()``` after calling ```imatlab_export_fig('fig2plotly')```?

**Comment by imatlab-helper** [comment]
To be clear, I don't believe a configuration of plotly has been necessary beyond the ```getplotlyoffline``` command. Try deleting the ```.config``` and ```.credentials``` files that get created in ```~/.plotly```. Perhaps there is something unusual in your configuration? The defaults that get initialized when those files don't exist should be correct.

The ```~/.plotly``` folder should only contain:
```.config```
```.credentials```
```plotlyjs/plotly-matlab-offline-bundle.js```

**Comment by twocs** [comment]
On Windows, I found that the plotly userhome is set to the %APPDATA% directory, not the user's home directory (reading header from `plotly\plotly_aux\getuserdir.m`). I'm not sure how `.config` and `.credentials` got into my home directory, but since I deleted them, now they don't exist.

I also found that some of the `try/catch` in the plotly matlab code is hiding error messages. For example, the error message `Oops! You must be signed in to initialize a plotlyfig object.` is hiding this message:

```
% check if credentials exist
if ~exist(plotly_credentials_file, 'file')
    error('Plotly:CredentialsNotFound',...
        ['It looks like you haven''t set up your plotly '...
        'account credentials yet.\nTo get started, save your '...
        'plotly username and API key by calling:\n'...
        '>>> saveplotlycredentials(username, api_key)\n\n'...
        'For more help, see https://plot.ly/MATLAB or contact '...
        'chris@plot.ly.']);
end
```

Is calling `saveplotlycredentials` something that I should do?

**Comment by imatlab-helper** [comment]
No, you should never need to call ```saveplotlycredentials()```. This should all be short-circuited.

It sounds like the plotly matlab API doesn't behave well in Windows when trying to use it in offline mode. I don't have access to a windows machine, so we may rely on you to help debugging here.

```fig2plotly``` should call ```plotlyfig``` and that should then call ```plotlyoffline``` here: https://github.com/plotly/MATLAB-Online/blob/master/plotly/plotlyfig.m#L445

Set a ```keyboard``` breakpoint just before that line. Are things getting that far or does it break earlier?

**Comment by imatlab-helper** [comment]
I just reread your initial post. Try commenting out lines 41 to 48 of ```plotlyfig```: https://github.com/plotly/MATLAB-Online/blob/master/plotly/plotlyfig.m#L41 and see what happens. 

I suspect what is going on here is that the function ```signin``` is being overloaded by something else in your path and thus not behaving as the plotly API expects it to. This is pretty bad on the plotly side since ```signin``` is a totally ambiguous function name and will commonly collide with other functions users might use.

**Comment by imatlab-helper** [comment]
It may also have something to do with this: plotly/MATLAB-online#98

**Comment by twocs** [comment]
I tried commenting out `signin` earlier and get different errors about PlotlyDomain. 
Before I look into more hacking of the plotly code I am wondering if you could tell me something.

Somehow the .config and .credentials are appearing in the user directory
when I first run `imatlab_export_fig('fig2plotly')` in the jupyter notebook. According to plotly, in a Windows system they should appear in the %APPDATA% directory, not the user directory (as I mentioned above).

Do you know how these files are getting created there (is it via the imatlab hook)? I tried poking around but can't find it. It doesn't make sense to me that the files are being created in the "wrong" directory.

**Comment by anntzer** [comment]
The only call imatlab does to any plotly-provided function is
```
fig2plotly(fig_handle, 'filename', '...', 'offline', true, 'open', false);
```
(https://github.com/imatlab/imatlab/blob/master/imatlab/resources/imatlab_export_fig.m#L39)

Can you try doing that from a clean plotly install (after getting the offline data) and see what happens?

**Comment by twocs** [comment]
Details about the problem (for posterity).

I found the cause of the incorrect folder creation. Every time that a new notebook with the imatlab kernel, the folder `~/.plotly` is created... because at `import plotly`, the Python plotly creates those files in the "wrong" folder for MATLAB offline plotly. This is because `plotly.py` regards `~/.plotly` as its directory rather than `MATLAB-offline` that uses `%APPDATA%/.plotly`. It seems hard coded as `PLOTLY_DIR = os.path.join(os.path.expanduser("~"), ".plotly")` in `https://github.com/plotly/plotly.py/blob/b35979e4016a668d949954718dc5dfc44b61ea8c/plotly/files.py`

Because of the call to `signin` occurring before plotly checking the "offline" flag, the credentials _must_ exist. In Linux/Mac, if the credentials don't exist, blanks are created in the "correct" directory. (There is a bug in plotly with the error hiding for missing credentials, but I'd say work around it rather than resolve it. Plotly also neglects to close the credentials file, so it cannot be manually edited until the notebook is closed.)

**Comment by twocs** [comment]
The workaround for Windows is to run this command once, either from MATLAB or in the Jupyter notebook with MATLAB kernel: `saveplotlycredentials('', '');`

This will overwrite any existing plotly credentials, so should only be done if the credentials aren't there. I think this could be included in the install instructions for Windows just before the instructions for `getplotlyoffline('https://cdn.plot.ly/plotly-latest.min.js')` or as a parenthetical section for Windows OS.

The other problem I found is that the file must be saved in the `pwd` for the notebook, i.e. not a custom path. So before saving a plot, I also call `cd(<target_dir>)`

Summary: If on Windows, call `saveplotlycredentials('', '');`

**Comment by anntzer** [comment]
Can you open an issue with plotly to try to get the issue resolved on their side (either Python or MATLAB)?  I can then just require a minimum version number (on either package).

> The other problem I found is that the file must be saved in the pwd for the notebook, i.e. not a custom path. So before saving a plot, I also call cd(<target_dir>)

I don't get at all what you mean?

**Comment by twocs** [comment]
The image file is saved as follows:  

> MATLAB-Online/blob/master/plotly/plotly_offline_aux/plotlyoffline.m lines 75-78
```
    % save the html file in the working directory
    plotly_offline_file = fullfile(pwd, html_filename); 
    file_id = fopen(plotly_offline_file, 'w');
    fprintf(file_id, offline_script); 
    fclose(file_id); 
```
So I don't see any way to have the offline file saved in a directory of my choosing. Did I miss something?

**Comment by twocs** [comment]
> Can you open an issue with plotly to try to get the issue resolved on their side (either Python or MATLAB)? I can then just require a minimum version number (on either package).

I see there are several problems that were created when Plotly decided to allow offline mode.
1. The Python statement `import plotly` creates dummy .credential and .config files in `~/.plotly` even if offline mode will never use those files.
2. `MATLAB-online` requires .credential and .config even if offline mode is used. The location is  `%APPHOME%/.plotly` for Windows and `~/.plotly` otherwise.
3. `MATLAB-online` hides the `Plotly:CredentialsNotFound` error with `plotlyfigConstructor:notSignedIn`
4. `MATLAB-online` exposes functions with very common names like `signin`.

As long as plotly is installed with `plotlysetup.m` then `saveplotlycredentials.m` will get called and the dummy credentials will exist. Also, their documentation states that `saveplotlycredentials` should be run or set manually https://plot.ly/matlab/getting-started/#credentials so in a way, as long as the documentation is followed, the bug won't be encountered.

So, basically I'm not sure where this bug would be filed. Is it `plotly.py` or `MATLAB-Online`, or both? Also, if you don't like the extra line of instructions for Windows, I think it should be easy for me to patch imatlab so that it's automated. However, it seems a necessary files issue on the same level as running `getplotlyoffline('https://cdn.plot.ly/plotly-latest.min.js')` to copy the JavaScript files. Would it be okay to automatically fetch that if it didn't exist?

**Comment by anntzer** [comment]
> So I don't see any way to have the offline file saved in a directory of my choosing. Did I miss something?

Internally, imatlab temporarily changes the pwd to a temporary directory before invoking `imatlab_export_fig`.  The created html is loaded into the notebook, then the temporary directory is removed.  (This design choice is exactly due to the behavior of `fig2plotly` that you describe.)

If you want to invoke `fig2plotly`/`plotlyoffline` yourself directly, this is an issue independent of imatlab.  You can file an issue with the MATLAB repo asking for the capability you require.

> 1. The Python statement import plotly creates dummy .credential and .config files in ~/.plotly even if offline mode will never use those files.
> 2. MATLAB-online requires .credential and .config even if offline mode is used. The location is %APPHOME%/.plotly for Windows and ~/.plotly otherwise.

For 1) I would suggest filing issues both in the Python and MATLAB repos linking each other, asking that the directory be unified.
For 2) I would suggest filing an issue in the MATLAB repo.

**Comment by twocs** [comment]
Looking at https://github.com/imatlab/imatlab/blob/master/README.rst to see a nice place to add a note for the Windows install. Just wanted to clarify the numbered list of steps.

The "Inline Graphics" instructions says to:
`addpath(genpath(<Plotly MATLAB API path>))`

But it seems that we do this every time that we run the notebook. Any reason not to suggest doing it like this:
`addpath(genpath(<Plotly MATLAB API path>)); savepath;`

The output of Step 3 gave me a bit of an impression about `fig2plotly` because the fprintf output states to generate the graph a certain way. So would it be okay to add something to this effect? "It is not necessary to invoke `fig2plotly` for imatlab."

Can I add an extra number to the list of steps for Windows?

Made this request which could close this issue: 
https://github.com/imatlab/imatlab/pull/15

**Comment by anntzer** [comment]
> the fprintf output states to generate the graph a certain way

what output?

**Comment by twocs** [comment]
> > the fprintf output states to generate the graph a certain way
> what output?

This is what we see when we run step 3:

**getplotlyoffline('https://cdn.plot.ly/plotly-latest.min.js')**
```
Success! You can generate your first offline graph
using the 'offline' flag of fig2plotly as follows:

>> plot(1:10); fig2plotly(gcf, 'offline', true);
```

These instructions are not applicable when using plotly for an exporter.

**Comment by anntzer** [comment]
Ah, I see.  Clarified in #16, please have a look.
